source/common/stats/stat_merger.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "envoy/stats/store.h"
+
+#include "common/common/hash.h"
+#include "common/protobuf/protobuf.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Responsible for the sensible merging of two instances of the same stat from two different
+// (typically hot restart parent+child) Envoy processes.
+class StatMerger {
+public:
+  StatMerger(Stats::Store& target_store);
+
+  // Merge the values of stats_proto into stats_store. Counters are always straightforward
+  // addition, while gauges default to addition but have exceptions.
+  void mergeStats(const Protobuf::Map<std::string, uint64_t>& counter_deltas,
+                  const Protobuf::Map<std::string, uint64_t>& gauges);
+
+  // TODO(fredlas) add void verifyCombineLogicSpecified(absl::string_view gauge_name), to
+  // be called at gauge allocation, to ensure (with an ASSERT) that anyone adding a new stat
+  // will be forced to come across this code and explicitly specify combination logic.
+  //
+  // OR,
+  // switch from the combination logic table to requiring the stat macro declarations themselves
+  // to indicate the logic.
+
+  // Returns true if the parent's value can be added in, false if we should do nothing.
+  static bool shouldImport(Gauge& gauge, const std::string& gauge_name);
+
+private:
+  void mergeCounters(const Protobuf::Map<std::string, uint64_t>& counter_deltas);
+  void mergeGauges(const Protobuf::Map<std::string, uint64_t>& gauges);
+  ConstCharStarHashMap<uint64_t> parent_gauge_values_;
```

Comment:
Use StatNameHashMap<uint64_t> now that the symbol-table rep is merged in.

Reply:
> is it the case that there are few enough gauges that, when it comes to gauges, we can just skip all the memory-saving fancy stat name stuff and still be safe?

I think it's fine to use more memory for now and get something functionally correct. :) But yeah I'm not too worried about gauges there are not that many of them. IMO we can live with full string storage or the symbol table variant, and I don't really even think pagination is that big of a deal but we can add that later. I'm most concerned about the leaking as an initial follow up.

Reply:
I suppose it's probably not going to be excessively expensive, but it seems pretty trivial to use the StatName as the key of the map, as that pattern is already established in thread_local_store.h.

Reply:
Reverted. Using that type sounds good for a followup.

Reply:
Haha ok, Josh came over and showed me what he was thinking of for the StatNameHashMap type. Very small change, and hopefully should still work.